Partitioner Hook for Custom Runtime Estimation

basilwong opened the pull request:
Summary:
## Summary

This diff adds a hook in PyTorch's partitioner to support **custom callable runtime estimators**.

### Motivation

The existing `RUNTIME_MODE` parameter only accepts predefined string modes (e.g., `"flops"`). To enable custom runtime estimation strategies—such as history-based estimation from MAST traces—we need the partitioner to accept user-provided callables.

Test Plan:
### E2E Testing

OmniFM LWB (96 B200; AC budget 0.3)
- Baseline + Custom Kernels + FLOPS Runtime Estimation: aps-LWB_omnifmv4_1124_autoac_flop-49be42744e 164k p90 QPS (+5%), 90% peak memory
- Baseline + Custom Kernels + History Based(**Custom**) Runtime Estimation: aps-LWB_omnifmv4_1124_autoac_lookup-aca6abaaa5 167k p90 QPS (+7%), 91% peak memory

Injecting custom runtime estimation, improves QPS results by ~2%.

Differential Revision: D88714606

basilwong commented:
@pytorchbot label "topic: not user facing"

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -2461,6 +2461,10 @@ def materialize_arg(x):
         return max(counted_flops, 1)
+
+    elif callable(RUNTIME_MODE):
+        return RUNTIME_MODE(node)
```
Making this a callable will probably break AOTAutograd caching. Today, the cache assumes that the corresponding config is a string constant that it can compute a cache key for: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/config.py#L175


If you want this config to optionally be a callable, then in order to make it cache-friendly you'll need to go through a similar route to our other "function config objects" - give it a specific type, and ensure that type provides a hash value that we can use to reliability compute a cache key. An inductor example for it is here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/config.py#L272C62-L272C81

basilwong replied:
Thanks for the heads up @bdhirsh. I've drafted the following pr based on your suggestions here: https://github.com/pytorch/pytorch/pull/171601

Let's continue iterating there.

basilwong committed:
Partitioner Hook for Custom Runtime Estimation (#170739)

Summary:

## Summary

This diff adds a hook in PyTorch's partitioner to support **custom callable runtime estimators**.

### Motivation

The existing `RUNTIME_MODE` parameter only accepts predefined string modes (e.g., `"flops"`). To enable custom runtime estimation strategies—such as history-based estimation from MAST traces—we need the partitioner to accept user-provided callables.

Test Plan:
### E2E Testing

OmniFM LWB (96 B200; AC budget 0.3)
- Baseline + Custom Kernels + FLOPS Runtime Estimation: aps-LWB_omnifmv4_1124_autoac_flop-49be42744e 164k p90 QPS (+5%), 90% peak memory
- Baseline + Custom Kernels + History Based(**Custom**) Runtime Estimation: aps-LWB_omnifmv4_1124_autoac_lookup-aca6abaaa5 167k p90 QPS (+7%), 91% peak memory

Injecting custom runtime estimation, improves QPS results by ~2%.

Reviewed By: avicizhu

Differential Revision: D88714606

atalman commented:
Please do not land this PR. The change is being reverted internally

atalman reviewed (approved):
lgtm
